antsibull/write_docs.py

```diff
@@ -0,0 +1,107 @@
+# coding: utf-8
+# Author: Toshio Kuratomi <tkuratom@redhat.com>
+# License: GPLv3+
+# Copyright: Ansible Project, 2020
+"""Output documentation."""
+
+import asyncio
+import os.path
+import typing as t
+
+import aiofiles
+import asyncio_pool
+from jinja2 import Template
+
+from .jinja2.environment import doc_environment
+from .constants import THREAD_MAX
+
+
+#: Mapping of plugins to nonfatal errors.  This is the type to use when accepting the plugin.
+#: The mapping is of plugin_type: plugin_name: [error_msgs]
+PluginErrorsT = t.Mapping[str, t.Mapping[str, t.Sequence[str]]]
+
+
+async def write_rst(plugin_name: str, plugin_type: str, plugin_record: t.Dict[str, t.Any],
+                    nonfatal_errors: PluginErrorsT, plugin_tmpl: Template, error_tmpl: Template,
+                    dest_dir: str) -> None:
+    """
+    Write the rst page for one plugin.
+
+    :arg plugin_name: FQCN for the plugin.
+    :arg plugin_type: The type of the plugin.  (module, inventory, etc)
+    :arg plugin_record: The record for the plugin.  doc, examples, and return are the
+        toplevel fields.
+    :arg nonfatal_errors: Mapping of plugin to any nonfatal errors that will be displayed in place
+        of some or all of the docs
+    :arg plugin_tmpl: Template for the plugin.
+    :arg error_tmpl: Template to use when there wasn't enough documentation for the plugin.
+    :arg dest_dir: Destination directory for the plugin data.  For instance,
+        :file:`ansible-checkout/docs/docsite/rst/`.  The directory structure underneath this
+        directory will be created if needed.
+    """
+    # plugin_record['doc']['name'] has the short name for the plugin.
+    # So collection_name for a plugin from a collection is the part in front of that.
+    try:
+        namespace, collection, plugin_short_name = plugin_name.split('.', 2)
+    except ValueError:
+        # Ansible doesn't give itself any names
+        namespace = 'ansible'
+        collection = 'builtins'
```

Comment:
```suggestion
        collection = 'builtin'
```
The synthetic collection included in ansible-base is called `ansible.builtin`, not `ansibe.builtins`. (If you grep for it, you'll find

Reply:
<nod> Yeah, I saw this too.  I've got the same simple fix queued, but I wondered whether I should modify the schemas to add ansible builtin to any DOCUMENTATION entry that only has a short name and then removing this code.

Reply:
Okay, after talking with nitzmahone about how the collection loader works, I don't think that the schema is the right place anymore.  The reason is theoretical but I think it's valid.

Modules in `ANSIBLE_LIBRARY` (and same for the other plugin paths) are actually unknown to the collection loader.  So to ansible, the plugins shipped with ansible-base are part of the synthetic ansible-builtin collection but the ones in ANSIBLE_LIBRARY path are known to not exist in a collection at all.  The ansible-doc parser is not passing that information on to us.

If we normalized this in the schema, then we'd falsely be adding `ANSIBLE_LIBRARY` plugins into `ansible.builtin`. Since the idea for the schema is that it could be easily reused by other parties, I don't think we should perform this kludge there.  I think the right place to  fix up the plugin names is in the backend parser.  Until we write our own, we have to rely on what ansible-doc does.  what  ithink that means is we should put the kludge into the ansible_doc backend.  After we get the output back from ansible_doc, change the plugin names that need them to include ansible.builtins there.

Reply:
This strategy has now been implemented https://github.com/ansible-community/antsibull/pull/5/files#diff-d0fd94da5008596d8d48e47df3c1e89eR108